modules/phase_field/src/kernels/TWSDoubleObstacle.C

```diff
@@ -0,0 +1,35 @@
+#include "TWSDoubleObstacle.h"
+
+registerMooseObject("PhaseFieldApp", TWSDoubleObstacle);
+
+InputParameters
+TWSDoubleObstacle::validParams()
+{
+  InputParameters params = ADKernelValue::validParams();
+  params.addRequiredParam<std::string>("sigma", "the interface energy used in the kernel");
+  params.addRequiredParam<std::string>("eta", "the interface width used in the kernel");
+  params.addRequiredParam<std::string>("delta_g", "the driving force used in the kernel");
+  params.addClassDescription("traveling wave kenel for double obstacle potential,"
+                             "-gamma * (phi - 0.5) - sqrt(phi * (1 - phi)) * m.");
+  return params;
+}
+
+TWSDoubleObstacle::TWSDoubleObstacle(const InputParameters & parameters)
+  : ADKernelValue(parameters),
+    _sigma_name(getParam<std::string>("sigma")),
+    _sigma(getADMaterialProperty<Real>(_sigma_name)),
+    _eta_name(getParam<std::string>("eta")),
+    _eta(getADMaterialProperty<Real>(_eta_name)),
+    _delta_g_name(getParam<std::string>("delta_g")),
+    _delta_g(getADMaterialProperty<Real>(_delta_g_name))
+{
+}
+
+ADReal
+TWSDoubleObstacle::precomputeQpResidual()
+{
+  auto value = -8.0 * _sigma[_qp] / _eta[_qp] * (_u[_qp] - 0.5);
+  if (_u[_qp] * (1.0 - _u[_qp]) > 1.0e-16)
+    value -= -8.0 / _pi / _delta_g[_qp] * std::sqrt(_u[_qp] * (1.0 - _u[_qp]));
+  return value;
```

Comment:
this has a pretty straight forward Jacobian, so I'd probably just implement this as non-AD...

Reply:
I have change this class to non-AD version. I wondered what are the advantages of AD and non-AD respectively, and how to choose AD or non-AD?

Reply:
Sorry, because I am not familiar with pull request, I have submitted many changes. Now, I have completed all the changes.

Reply:
@Mascsu , AD can be very useful when the Jacobian cannot be easily written down. For example in multiphysics coupling, where material properties can depend on many variables, and in large deformation problems, where many quantities implicitly depend on the displacements.